sync_cleared_for_campus.R

```diff
@@ -0,0 +1,179 @@
+#load_dot_env('prod.env')
+
+# get_case_records suitability is questionable
+case_records <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
+            token = Sys.getenv("CASE_TOKEN"),
+            events = c('case_investigation_arm_1'),
+            guess_max = 3000)$data %>%
+    mutate(case_emp_ufid = as.character(case_emp_ufid)) %>% # type change for dplyr
+    filter(is.na(redcap_repeat_instrument) & !is.na(case_emp_ufid)) %>%
+    mutate(cleared_for_campus_timestamp = as.character(cleared_for_campus_timestamp)) %>%
+    select(record_id, redcap_event_name, case_emp_ufid, contains("cleared_for_campus"))
+
+# get_contact_records is not suitable
+contact_records <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
+                                               token = Sys.getenv("CONTACT_TOKEN"),
+                                               events = c('contact_tracing_arm_1'),
+                                               guess_max = 3000)$data %>%
+    mutate(cleared_for_campus_timestamp = as.character(cleared_for_campus_timestamp))
+
+# FIXME: not using oneshot results in type coercion error
+result_fields <- c("record_id", "cleared_for_campus", "cleared_for_campus_timestamp")
+result_records <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
+        token = Sys.getenv("RESULT_TOKEN"),
+        fields = result_fields,
+        guess_max = 3000)$data %>%
+    mutate(cleared_for_campus_timestamp = as.character(cleared_for_campus_timestamp))
+
+message <- ""
+
+# TODO: move this function to the appropriate functions file after refactor
+sync_fields_by_timestamp <- function(dataset_a, join_field_in_a, dataset_b, join_field_in_b, timestamp_field, field_to_sync = '') {
+    # this function sets fields_to_sync to those associated with the latest timestamp_field
+
+    dataset <- dataset_a %>%
+        inner_join(dataset_b, by = setNames(nm=join_field_in_a, join_field_in_b), suffix = c('.a', '.b'))
+    # setNames creates c(join_field_in_a = join_field_in_b) using the _values_ of the join_field_in_<> variables
+
+
+    # hack variable name strings for dplyr use
+    timestamp_field.a = paste0(timestamp_field, '.a')
+    timestamp_field.b = paste0(timestamp_field, '.b')
+    # create symbolic variables from the _values_ of variables
+    # i.e. from foo = String bar create Object bar
+    ta = sym(timestamp_field.a)
+    tb = sym(timestamp_field.b)
+
+    field_to_sync.a = paste0(field_to_sync, '.a')
+    field_to_sync.b = paste0(field_to_sync, '.b')
+    fa = sym(field_to_sync.a)
+    fb = sym(field_to_sync.b)
+
+    # use latest timestamp as timestamp_field with no suffix
+    # and update field_to_sync to match field with latest timestamp
+    # using symbols here prevents the logical operator performing string literal comparison
+    synced_dataset <- dataset %>%
+        filter( (!!ta != !!tb) | xor(is.na(!!ta), is.na(!!tb)) ) %>%
+        mutate(!!timestamp_field := max(!!ta, !!tb, na.rm = T)) %>%
```

Comment:
```suggestion
        mutate(!!timestamp_field := case_when(
            ymd_hms(!!ta) > ymd_hms(!!tb) ~ !!ta,
            ymd_hms(!!tb) > ymd_hms(!!ta) ~ !!tb,
            is.na(!!ta) ~ !!tb,
            is.na(!!tb) ~ !!ta,
            TRUE ~ !!ta # default to Result in event of identical timestamps or unknown case
            )
        ) %>%
```

hunk 1 of commit 17919af was bad. The previous hunk appears to be perfect.

Reply:
Oh no, I really liked Laurence's shortening of it so I guess I didn't test it closely enough.